File: benchmark/binary/x86-64/window/small/source/golang/structs_and_pointers.go
package main

import "fmt"

// Record struct
type Record struct {
	ID    int
	Value int
	Name  string
}

// Process record
func (r *Record) Process() int {
	return r.ID + r.Value
}

// Node for linked list
type Node struct {
	Data int
	Next *Node
	Prev *Node
}

// LinkedList implementation
type LinkedList struct {
	Head *Node
}

// Insert at head
func (ll *LinkedList) InsertAtHead(data int) {
	newNode := &Node{Data: data}
	if ll.Head != nil {
		newNode.Next = ll.Head
		ll.Head.Prev = newNode
	}
	ll.Head = newNode
}

// Sum all nodes
func (ll *LinkedList) Sum() int {
	sum := 0
	current := ll.Head
	for current != nil {
		sum += current.Data
		current = current.Next
	}
	return sum
}

// Get length
func (ll *LinkedList) Length() int {
	count := 0
	current := ll.Head
	for current != nil {
		count++
		current = current.Next
	}
	return count
}

// Point structure with generics simulation
type Point struct {
	X, Y, Z float64
}

// Distance from origin
func (p *Point) DistanceFromOrigin() float64 {
	return p.X*p.X + p.Y*p.Y + p.Z*p.Z
}

// Translate point
func (p *Point) Translate(dx, dy, dz float64) {
	p.X += dx
	p.Y += dy
	p.Z += dz
}

// RecordManager
type RecordManager struct {
	Records []Record
}

// Add record
func (rm *RecordManager) AddRecord(rec Record) {
	rm.Records = append(rm.Records, rec)
}

// Get total value
func (rm *RecordManager) GetTotalValue() int {
	total := 0
	for _, rec := range rm.Records {
		total += rec.Value
	}
	return total
}

// Count records
func (rm *RecordManager) Count() int {
	return len(rm.Records)
}

// Pointer arithmetic with slices
type VectorOps struct {
	Data []int
}

// Sum via slice iteration
func (vo *VectorOps) Sum() int {
	sum := 0
	for _, val := range vo.Data {
		sum += val
	}
	return sum
}

// Modify via reference
func (vo *VectorOps) ModifyAt(index, value int) {
	if index >= 0 && index < len(vo.Data) {
		vo.Data[index] = value
	}
}

func main() {
	// Test Record
	rec1 := &Record{ID: 1, Value: 100, Name: "Alice"}
	rec2 := &Record{ID: 2, Value: 200, Name: "Bob"}
	calc1 := rec1.Process()
	calc2 := rec2.Process()

	// Test LinkedList
	list := &LinkedList{}
	list.InsertAtHead(10)
	list.InsertAtHead(20)
	list.InsertAtHead(30)
	list.InsertAtHead(40)
	listSum := list.Sum()
	listLen := list.Length()

	// Test Point
	point1 := &Point{X: 1.0, Y: 2.0, Z: 3.0}
	dist := point1.DistanceFromOrigin()
	point1.Translate(1.0, 1.0, 1.0)

	// Test RecordManager
	manager := &RecordManager{}
	manager.AddRecord(*rec1)
	manager.AddRecord(*rec2)
	totalValue := manager.GetTotalValue()
	recCount := manager.Count()

	// Test VectorOps
	arr := []int{1, 2, 3, 4, 5}
	vecOps := &VectorOps{Data: arr}
	vecSum := vecOps.Sum()
	vecOps.ModifyAt(0, 10)

	fmt.Printf("Structs and pointers test: %d %d %d %d %d %d\n",
		calc1, calc2, listSum, totalValue, int(dist), vecSum)
	fmt.Printf("List length: %d, Record count: %d\n", listLen, recCount)
}

File: benchmark/binary/x86-64/window/small/source/c/structs_and_pointers.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Structure definitions for testing
typedef struct {
    int id;
    int value;
    char name[32];
} Record;

typedef struct Node {
    int data;
    struct Node *next;
    struct Node *prev;
} Node;

typedef struct {
    int x;
    int y;
    int z;
} Point3D;

// Function with struct parameter
int process_record(Record *rec) {
    if (rec == NULL) {
        return -1;
    }
    return rec->id + rec->value;
}

// Function that creates and manipulates structs
Record* create_record(int id, int value, const char *name) {
    Record *rec = (Record *)malloc(sizeof(Record));
    if (rec == NULL) {
        return NULL;
    }
    
    rec->id = id;
    rec->value = value;
    strncpy(rec->name, name, sizeof(rec->name) - 1);
    rec->name[sizeof(rec->name) - 1] = '\0';
    
    return rec;
}

// Linked list operations
Node* create_node(int data) {
    Node *node = (Node *)malloc(sizeof(Node));
    if (node != NULL) {
        node->data = data;
        node->next = NULL;
        node->prev = NULL;
    }
    return node;
}

// Insert node at beginning of list
Node* insert_at_head(Node *head, int data) {
    Node *new_node = create_node(data);
    if (new_node == NULL) {
        return head;
    }
    
    if (head != NULL) {
        new_node->next = head;
        head->prev = new_node;
    }
    
    return new_node;
}

// Sum all values in linked list
int sum_list(Node *head) {
    int sum = 0;
    Node *current = head;
    
    while (current != NULL) {
        sum += current->data;
        current = current->next;
    }
    
    return sum;
}

// Free linked list
void free_list(Node *head) {
    Node *current = head;
    while (current != NULL) {
        Node *next = current->next;
        free(current);
        current = next;
    }
}

// Function with array of pointers
int process_records(Record **records, int count) {
    int total = 0;
    
    for (int i = 0; i < count; i++) {
        if (records[i] != NULL) {
            total += records[i]->value;
        }
    }
    
    return total;
}

// Pointer arithmetic
int sum_array_via_pointers(int *arr, int len) {
    int sum = 0;
    int *ptr = arr;
    int *end = arr + len;
    
    while (ptr < end) {
        sum += *ptr;
        ptr++;
    }
    
    return sum;
}

// Double pointer manipulation
void modify_via_double_pointer(int **ptr_to_ptr, int value) {
    if (ptr_to_ptr != NULL && *ptr_to_ptr != NULL) {
        **ptr_to_ptr = value;
    }
}

// Main entry point
int main() {
    // Create records
    Record *rec1 = create_record(1, 100, "Alice");
    Record *rec2 = create_record(2, 200, "Bob");
    Record *rec3 = create_record(3, 300, "Charlie");
    
    // Array of record pointers
    Record *rec_array[] = {rec1, rec2, rec3};
    int total_records = process_records(rec_array, 3);
    
    // Linked list operations
    Node *list = NULL;
    list = insert_at_head(list, 10);
    list = insert_at_head(list, 20);
    list = insert_at_head(list, 30);
    list = insert_at_head(list, 40);
    
    int list_sum = sum_list(list);
    
    // Pointer arithmetic
    int arr[] = {1, 2, 3, 4, 5};
    int arr_sum = sum_array_via_pointers(arr, 5);
    
    // Double pointer
    int value = 42;
    int *ptr = &value;
    modify_via_double_pointer(&ptr, 100);
    
    // Cleanup
    free(rec1);
    free(rec2);
    free(rec3);
    free_list(list);
    
    printf("Total: %d %d %d\n", total_records, list_sum, arr_sum);
    
    return 0;
}
